bdai_ros2_wrappers/bdai_ros2_wrappers/callables.py

```diff
@@ -163,3 +165,207 @@ def asynchronous(self, *args: Any, **kwargs: Any) -> Any:
+            """Invoke method synchronously."""
+            return self.body.synchronous(*args, **kwargs)
+
+        def asynchronous(self, *args: Any, **kwargs: Any) -> Any:
+            """Invoke method asynchronously."""
+            return self.body.asynchronous(*args, **kwargs)
+
+    def __init__(self, prototype: Callable, transitional: bool) -> None:
+        """Initializes the generalized method.
+
+        Args:
+            prototype: method prototype, usually just a signature but
+            may also be used as an overload for convenience (iff the
+            function type matches the missing overload).
+            transitional: a transitional method will stick to its
+            prototype for default invocations, simplifying the
+            adoption of generalized methods in existing codebases.
+        """
+        self.prototype = prototype
+        self.transitional = transitional
+        self.synchronous_overload: Optional[Callable] = None
+        self.asynchronous_overload: Optional[Callable] = None
+
+    def sync_overload(self, func: Callable) -> Callable:
+        """Register `func` as this method synchronous overload."""
+        if self.synchronous_overload is not None:
+            raise RuntimeError("cannot redefine synchronous overload")
+        self.synchronous_overload = func
+        return func
+
+    def async_overload(self, func: Callable) -> Callable:
+        """Register `func` as this method asynchronous overload."""
+        if self.asynchronous_overload is not None:
+            raise RuntimeError("cannot redefine asynchronous overload")
+        self.asynchronous_overload = func
+        return func
+
+    def __set_name__(self, owner: Type, name: str) -> None:
+        self.__attribute_name = f"__{name}_method"
+        setattr(owner, self.__attribute_name, GeneralizedMethod.Unbound(self))
+
+    def rebind(self, instance: Any, body: GeneralizedCallable) -> None:
+        """Change this method's `body` for the given `instance`."""
+        default_callable: Optional[Callable] = None
+        if self.transitional:
+            default_callable = self.prototype.__get__(instance)
+        bound_method = GeneralizedMethod.Bound(body, default_callable)
+        setattr(instance, self.__attribute_name, bound_method)
+
+    @overload
+    def __get__(self, instance: Literal[None], owner: Optional[Type] = ...) -> "GeneralizedMethod":
+        ...
+
+    @overload
+    def __get__(self, instance: Any, owner: Optional[Type] = ...) -> "GeneralizedMethod.Bound":
+        ...
+
+    def __get__(
+        self,
+        instance: Optional[Any],
+        owner: Optional[Type] = None,
+    ) -> Union["GeneralizedMethod", "GeneralizedMethod.Bound"]:
+        if instance is None:
+            return self
+        return getattr(instance, self.__attribute_name)
+
+
+@overload
+def generalizedmethod(func: Callable, *, transitional: bool = ...) -> GeneralizedMethod:
```

Comment:
```suggestion
def generalized_method(func: Callable, *, transitional: bool = ...) -> GeneralizedMethod:
```

Reply:
Done in https://github.com/bdaiinstitute/ros_utilities/pull/93/commits/8fbf659084005bf7dfe415aac178b8c04c1a9a70.